torch/csrc/jit/passes/onnx/scalar_type_analysis.cpp

```diff
@@ -162,36 +191,22 @@ static c10::optional<c10::ScalarType> InferExpectedScalarType(const Node* n) {
           auto tensor = input->node()->t(attr::value);
           auto rank = tensor.dim();
           auto scalar_type = tensor.scalar_type();
-          // Mimic PyTorch scalar type promotion logic
-          // from https://github.com/pytorch/pytorch/issues/9515
-          // Quoting:
-          //    A Tensor is a considered a "wrapped number" if it is
-          //    auto-wrapped from a C++ or Python number type. Integer types are
-          //    wrapped as 0-dim int64 tensors and floating-point types are
-          //    wrapped as 0-dim double tensors.
+
           if (rank == 0) {
-            auto default_scalar_type =
-                at::typeMetaToScalarType(at::get_default_dtype());
-            switch (scalar_type) {
-              case at::kDouble:
-                // floating-point numbers wrapped as double tensors are
-                // considered to have default type, instead of double.
-                typesFromScalars.emplace_back(default_scalar_type);
-                break;
-              case at::kLong:
-              case at::kBool:
-                // bool and integer numbers remain the same type.
-                typesFromScalars.emplace_back(scalar_type);
-                break;
-              default:
-                // other types are not from wrapped numbers,
-                // track them as types from tensors.
-                typesFromTensors.emplace_back(scalar_type);
-                break;
-            }
+            emplace_type_from_scalar(scalar_type);
           } else {
             typesFromTensors.emplace_back(scalar_type);
           }
+        } else if (nkind == prim::Param) {
```

Comment:
Could you briefly explain why below logic only applies on outputs from prim::Param nodes?

Reply:
It's for converting graph input scalar (e.g., int) to tensor.

Reply:
But why only graph input, not other values (handled by the last branch on L210), has this logic? I'd recommend commenting the reasoning.

Reply:
No exporter support scalar as graph input before. If you enable this for other operators, those existing exporters may fail.

Reply:
This PR is making regression on #85316. I am still working on the issue, any insight would be appreciated.

Reply:
Specifically, I found it turns Double to FLOAT (or adding Cast nodes)